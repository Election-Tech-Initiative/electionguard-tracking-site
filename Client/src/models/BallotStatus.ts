export class BallotStatus {
  public trackingId?: string;
  public status?: number;
  public location?: string;
  public approximateCastTime?: Date;
  public details?: string[];

  public matchedSegment?: string;
  public unmatchedSegment?: string;
}
