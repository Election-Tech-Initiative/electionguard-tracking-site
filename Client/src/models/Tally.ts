export class Tally {
  public selectionId?: string;
  public name?: string;
  public description?: string;
  public party?: string;
  public voteCount?: number;
}
