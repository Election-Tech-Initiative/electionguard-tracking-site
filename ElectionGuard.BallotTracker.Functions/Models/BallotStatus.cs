using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace ElectionGuard.BallotTracker.Functions.Models
{
    public class BallotStatus
    {
        public string TrackingId { get; set; }

        public BallotCountStatus Status { get; set; }

        public string Location { get; set; }

        public DateTime? ApproximateCastTime { get; set; }

        public IEnumerable<string> Details { get; set; }
    }
}
