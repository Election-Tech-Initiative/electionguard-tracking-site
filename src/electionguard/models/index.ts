export * from './election_object_base';
export * from './election';
export * from './tally';
