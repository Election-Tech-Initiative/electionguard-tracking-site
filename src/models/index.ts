export * from './election';
export * from './internationalizedText';
export * from './tally';
export * from './tracking';
