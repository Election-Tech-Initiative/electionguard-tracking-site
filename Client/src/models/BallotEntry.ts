import { Tally } from './Tally';

export class BallotEntry {
  public raceId?: string;
  public name?: string;
  public description?: string;
  public tallies?: Tally[];
}
