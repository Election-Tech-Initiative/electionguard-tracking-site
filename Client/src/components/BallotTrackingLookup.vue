<template>
  <div id="lookup" onclick="">
    <label class="label">{{ $t('trackerSearchLabel') }}</label>
    <div class="control" :class="{'is-loading': isLoading}">
      <vue-autosuggest
        v-model="query"
        @input="lookupBallots"
        @selected="onSelected"
        :get-suggestion-value="getSuggestionValue"
        :inputProps="{
          id: 'autosuggest__input',
          placeholder: $t('trackerSearchInputPlaceholder'),
          class: 'input',
        }"
        :limit="limit"
        :suggestions="suggestions">
        <template slot="before-input">
          <a v-if="!isLoading && query.length > 1" class="delete" @click.prevent="reset"></a>
        </template>
        <template slot="before-suggestions">
          <h6 class="subtitle is-6">{{ $t('trackerSearchAutocompleteLabel') }}</h6>
        </template>
        <template slot-scope="{suggestion}">
          <div>
            <span class="has-text-weight-semibold">{{suggestion.item.matchedSegment}}</span><span>{{suggestion.item.unmatchedSegment}}</span>
          </div>
        </template>
        <template slot="after-suggestions" v-if="ballots.length > limit">
          <h6 class="subtitle is-6 after-suggestions">{{ $t('trackerSearchMoreResultsMessage', { moreResultsCount: ballots.length - limit}) }} </h6>
        </template>
      </vue-autosuggest>
    </div>
  </div>
</template>

<script lang="ts">
import { Component, Prop, Vue, Mixins } from 'vue-property-decorator';
import { VueAutosuggest } from 'vue-autosuggest';

import { BallotStatus } from '@/models/BallotStatus';

import BallotTrackingApiMixin from '@/mixins/BallotTrackingApi';

@Component({
  components: {
    VueAutosuggest
  },
})
export default class BallotTrackingLookup extends Mixins(BallotTrackingApiMixin) {
  public ballots: BallotStatus[] = [];
  public query: string = '';

  private previousQuery = '';
  private readonly MinimumSearchLength = 3;

  public getSuggestionValue(suggestion: { item: BallotStatus }): string | undefined {
    return suggestion.item.trackingId;
  }

  public get limit() {
    const isMobile = navigator.userAgent.match(
      /Mobile|Android|webOS|iPhone|iPad|iPod|BlackBerry|BB|PlayBook|IEMobile|Windows Phone|Kindle|Silk|Opera Mini/i)
      !== null;
    return isMobile ? 5 : 10;
  }

  public async lookupBallots(query: string): Promise<void> {
    const pattern = this.getPatternString(query);
    if (this.isTooFewCharsToSearch(pattern)) {
      this.ballots = [];
      this.previousQuery = '';
    } else {
      if (this.shouldLookup(pattern)) {
        this.ballots = await this.queryApiForMatchingBallots(pattern);
        this.ballots.forEach((b) => {
          this.updateTrackingSegments(b, pattern);
        });
      } else {
        // TODO: If needed, optimize this so that we still keep the previously
        // returned entries in memory and reload them if the the pattern is
        // requeried (someone hits the backspace)
        this.ballots = this.ballots.filter((b) => {
          const matched = this.getPatternString(b.trackingId).startsWith(pattern);
          if (matched) {
            this.updateTrackingSegments(b, pattern);
          }
          return matched;
        });
      }

      // if no ballots found, add in the entry take user to the not found modal
      if (this.ballots.length === 0) {
        const notFoundStatus: BallotStatus = {
          trackingId: query,
          status: 404,
          matchedSegment: this.$t('trackerSearchNotFound').toString(),
        };
        this.ballots = [notFoundStatus];
      }

      this.previousQuery = pattern;
    }
  }

  public onSelected(suggestion: { item: BallotStatus }): void {
    if (!suggestion || !suggestion.item) {
      return;
    }

    const { item: ballot } = suggestion;
    if (ballot.trackingId) {
      this.$emit('selected', ballot);
      // after emitting selected up to the parent component to handle,
      // re-run the lookup again so that when modal closes, only current option exists
      this.lookupBallots(ballot.trackingId);
    }
  }

  public reset(): void {
    this.ballots = [];
    this.query = '';
  }

  public get suggestions() {
    return [{ data: this.ballots } ];
  }

  private isTooFewCharsToSearch(pattern: string) {
    return pattern.length < this.MinimumSearchLength;
  }

  private shouldLookup(pattern: string): boolean {
    // TODO: Make sure that the previous query returned less than the max items
    // allowed to be returned from a lookup.  Thus, if the query was narrowed we may
    // be missing items that weren't returne din the first query.
    return !pattern.startsWith(this.previousQuery)
        || pattern.length < this.previousQuery.length
        || !this.previousQuery
        || this.previousQuery === '';
  }

  private updateTrackingSegments(ballot: BallotStatus, pattern: string): void {
    if (!ballot || !ballot.trackingId) {
      return;
    }

    let matchedCount = 0;
    let index = 0;
    while (matchedCount < pattern.length) {
      if (ballot.trackingId[index] !== ' ') {
        // Not a space, so increment matchedCount
        matchedCount += 1;
      }
      index += 1;
    }

    ballot.matchedSegment = ballot.trackingId.substr(0, index);
    if (index < ballot.trackingId.length) {
      ballot.unmatchedSegment = ballot.trackingId.substr(index);
    } else {
      ballot.unmatchedSegment = '';
    }
  }
}
</script>

<style lang="scss">
#lookup {
  .control.is-loading:after {
    font-size: 1.8rem;
  }
  
  a.delete {
    position: absolute;
    top: 1.1rem;
    right: 1rem;
    z-index: 1;
    padding: 0.8rem;
  }

  .input {
    padding: 2rem 3rem 2rem 1.7rem;
    font-size: 1rem;
    font-weight: bold;

    &.autosuggest__input--open {
      border-bottom-left-radius: 0;
      border-bottom-right-radius: 0;
    }
  }

  .autosuggest__results-container {
    background-color: $white;
    border: 1px solid $grey-lighter;
    border-radius: 0 0 0.5rem 0.5rem;
    margin-top: -0.5rem;

    h6 {
      margin: 1.5rem 1.5rem 0.8rem;

      &.after-suggestions {
        color: $primary;
        margin-top: 1rem;
        margin-bottom: 1.5rem;
      }
    }

    ul {
      list-style: none;
      margin: 0;

      li {
        padding: 1rem 1.5rem;

        &.autosuggest__results-item--highlighted {
          background-color: $primary;
          color: $white;
        }

        &:hover {
          cursor: pointer;
        }
      }
    }
  }

  button {
    margin-top: 1rem;
  }
}
</style>
