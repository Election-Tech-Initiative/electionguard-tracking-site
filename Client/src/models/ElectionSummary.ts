import { BallotEntry } from './BallotEntry';

export class ElectionSummary {
  public isComplete!: boolean;
  public ballotEntries?: BallotEntry[];
}
