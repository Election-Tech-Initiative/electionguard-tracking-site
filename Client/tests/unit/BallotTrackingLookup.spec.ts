import { mount } from '@vue/test-utils';

import BallotTrackingLookup from '@/components/BallotTrackingLookup.vue';
import { BallotStatus } from '@/models/BallotStatus';

const mockedBallots = [{
    approximateCastTime: new Date('July 31, 2019 12:30:00'),
    location: 'mockedLocation',
    status: 0,
    trackingId: 'mockedTrackingId',
}];

describe('BallotTrackingLookup.vue', () => {
    const getMountedWrapper = (ballotData: BallotStatus[]) => mount(BallotTrackingLookup, {
        mocks: {
            appSettings: {
                apiUrl: 'mocked-url',
            },
            $http: {
                get: async (): Promise<any> => Promise.resolve({
                    status: 200,
                    data: ballotData,
                }),
            },
            $t: (msg: string): string => msg
        },
    });

    it('renders', async () => {
        const wrapper = getMountedWrapper([]);
        expect(wrapper.element).toMatchSnapshot();
    });

    it('renders with not found', async () => {
        const wrapper = getMountedWrapper([]);
        await (wrapper.vm as any).lookupBallots('emptyquery');
        wrapper.find('input').trigger('click'); // opens the dropdown results
        expect(wrapper.element).toMatchSnapshot('with not found');
    });

    it('renders with a suggestion', async () => {
        const wrapper = getMountedWrapper(mockedBallots);
        await (wrapper.vm as any).lookupBallots('mocked');
        wrapper.find('input').trigger('click'); // opens the dropdown results
        expect(wrapper.element).toMatchSnapshot('with a suggestion');
    });

    it('emits selected ballot when onSelected', () => {
        const wrapper = getMountedWrapper(mockedBallots);
        (wrapper.vm as any).onSelected({ item: mockedBallots[0] });
        expect(wrapper.emitted().selected).toBeTruthy();
    });
});
