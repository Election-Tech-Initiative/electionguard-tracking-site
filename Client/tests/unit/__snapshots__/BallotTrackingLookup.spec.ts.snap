// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BallotTrackingLookup.vue renders 1`] = `
<div
  id="lookup"
  onclick=""
>
  <label
    class="label"
  >
    trackerSearchLabel
  </label>
   
  <div
    class="control"
  >
    <div
      id="autosuggest"
    >
      <input
        aria-activedescendant=""
        aria-autocomplete="list"
        aria-expanded="false"
        aria-haspopup="false"
        aria-owns="autosuggest__results"
        autocomplete="off"
        class="input"
        id="autosuggest__input"
        placeholder="trackerSearchInputPlaceholder"
        role="combobox"
        type="text"
      />
       
      <div
        class="autosuggest__results-container"
      >
        <!---->
      </div>
    </div>
  </div>
</div>
`;

exports[`BallotTrackingLookup.vue renders with a suggestion: with a suggestion 1`] = `
<div
  id="lookup"
  onclick=""
>
  <label
    class="label"
  >
    trackerSearchLabel
  </label>
   
  <div
    class="control"
  >
    <div
      id="autosuggest"
    >
      <input
        aria-activedescendant=""
        aria-autocomplete="list"
        aria-expanded="true"
        aria-haspopup="true"
        aria-owns="autosuggest__results"
        autocomplete="off"
        class="autosuggest__input--open input"
        id="autosuggest__input"
        placeholder="trackerSearchInputPlaceholder"
        role="combobox"
        type="text"
      />
       
      <div
        class="autosuggest__results-container"
      >
        <div
          aria-labelledby="autosuggest"
          class="autosuggest__results"
        >
          <h6
            class="subtitle is-6"
          >
            trackerSearchAutocompleteLabel
          </h6>
           
          <ul
            aria-labelledby="autosuggest"
            role="listbox"
          >
            <li
              class="autosuggest__results-item"
              data-section-name="default"
              data-suggestion-index="0"
              id="autosuggest__results-item--0"
              role="option"
            >
              <div>
                <span
                  class="has-text-weight-semibold"
                >
                  mocked
                </span>
                <span>
                  TrackingId
                </span>
              </div>
            </li>
          </ul>
           
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`BallotTrackingLookup.vue renders with not found: with not found 1`] = `
<div
  id="lookup"
  onclick=""
>
  <label
    class="label"
  >
    trackerSearchLabel
  </label>
   
  <div
    class="control"
  >
    <div
      id="autosuggest"
    >
      <input
        aria-activedescendant=""
        aria-autocomplete="list"
        aria-expanded="true"
        aria-haspopup="true"
        aria-owns="autosuggest__results"
        autocomplete="off"
        class="autosuggest__input--open input"
        id="autosuggest__input"
        placeholder="trackerSearchInputPlaceholder"
        role="combobox"
        type="text"
      />
       
      <div
        class="autosuggest__results-container"
      >
        <div
          aria-labelledby="autosuggest"
          class="autosuggest__results"
        >
          <h6
            class="subtitle is-6"
          >
            trackerSearchAutocompleteLabel
          </h6>
           
          <ul
            aria-labelledby="autosuggest"
            role="listbox"
          >
            <li
              class="autosuggest__results-item"
              data-section-name="default"
              data-suggestion-index="0"
              id="autosuggest__results-item--0"
              role="option"
            >
              <div>
                <span
                  class="has-text-weight-semibold"
                >
                  trackerSearchNotFound
                </span>
                <span>
                  
                </span>
              </div>
            </li>
          </ul>
           
        </div>
      </div>
    </div>
  </div>
</div>
`;
